gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+    ///
+    /// Encoded:
+    ///   - `a`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `b`: `ArkScale<Vec<G2Affine>>`.
+    #[codec(index = 0)]
+    MultiMillerLoop { a: Vec<u8>, b: Vec<u8> },
+
+    /// Request to pairing final exponentiation for *BLS12-381*.
+    ///
+    /// Encoded: `ArkScale<<Bls12_381::TargetField>`.
+    #[codec(index = 1)]
+    FinalExponentiation { f: Vec<u8> },
+
+    /// Request to multi scalar multiplication on *G1* for *BLS12-381*
+    ///
+    /// Encoded:
+    ///   - `bases`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `scalars`: `ArkScale<Vec<G1Config::ScalarField>>`.
+    #[codec(index = 2)]
+    MultiScalarMultiplicationG1 { bases: Vec<u8>, scalars: Vec<u8> },
+
+    /// Request to multi scalar multiplication on *G2* for *BLS12-381*
+    ///
+    /// Encoded:
+    ///   - `bases`: `ArkScale<Vec<G2Affine>>`.
+    ///   - `scalars`: `ArkScale<Vec<G2Config::ScalarField>>`.
+    #[codec(index = 3)]
+    MultiScalarMultiplicationG2 { bases: Vec<u8>, scalars: Vec<u8> },
+
+    /// Request to projective multiplication on *G1* for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `base`: `ArkScaleProjective<G1Projective>`.
+    ///   - `scalar`: `ArkScale<Vec<u64>>`.
+    #[codec(index = 4)]
+    ProjectiveMultiplicationG1 { base: Vec<u8>, scalar: Vec<u8> },
+
+    /// Request to projective multiplication on *G2* for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `base`: `ArkScaleProjective<G2Projective>`.
+    ///   - `scalar`: `ArkScale<Vec<u64>>`.
+    #[codec(index = 5)]
+    ProjectiveMultiplicationG2 { base: Vec<u8>, scalar: Vec<u8> },
+}
+
+/// The enumeration represents possible common errors for all requests.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum Error {
+    /// Failed to scale-decode the length of a `Vec<u8>`.
+    DecodeVecLength,
+    /// Failed to scale-decode bytes.
+    DecodeVecData,
+    /// Failed to decode the length of a `Vec<G(1,2)Affine>`.
+    DecodeItemCount,
+}
+
+/// The enumeration contains result to a request.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::From)]
+#[codec(crate = codec)]
+pub enum Response {
+    /// Common error.
+    #[from]
+    Error(Error),
+    /// Result of the multi Miller loop [`MultiMillerLoopResult`].
+    #[from]
+    MultiMillerLoop(MultiMillerLoopResult),
+    /// Result of the final exponentiation.
+    FinalExponentiation(Result<Vec<u8>, ()>),
+    /// Result of the multi scalar multiplication [`MultiScalarMultiplicationResult`].
+    #[from]
+    MultiScalarMultiplication(MultiScalarMultiplicationResult),
+    /// Result of the projective multiplication.
+    ProjectiveMultiplication(Result<Vec<u8>, ()>),
+}
+
+/// Result of the multi Miller loop computation.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum MultiMillerLoopResult {
+    /// Encoded: `ArkScale<Bls12_381::TargetField>`.
+    Ok(Vec<u8>),
```

Comment:
In my opinion returning a byte vec to the caller affects the usability of the contract by a lot (hint: not in a good way). If I were a user of this builtin actor, I'd be confused: how am I supposed to decode the bytes into the result of some paired crypto operation?
Same with encoding and decoding in other requests.
If dealing with `ArkScale<_>` struct prevents accurate gas estimation I'd sacrifice the precision in favour of usability.

Reply:
The answer is the same as for the other remark: it is a consequence of the `sp-crypto-ec-utils` API.

Reply:
Still, the purpose of this crate is to abstract away encoding details from users. Otherwise one might argue that gear programs API requires the `handle` payload as a vector so let the users build the payload as they like - why bother creating any helper crates like this one then?

Reply:
Sorry, but cannot agree. The purpose of the builtin is just to re-export `sp-crypto-ec-utils` API to the gear-programs and that API makes some constraints. Yep, we are able to make API less noisy but that would impact gas consumption.

Reply:
> Sorry, but cannot agree. The purpose of the builtin is just to re-export `sp-crypto-ec-utils` API to the gear-programs and that API makes some constraints. Yep, we are able to make API less noisy but that would impact gas consumption.

How would this impact gas consumption if the same set of operations is performed in the end? The user who engages the calling contract will pay for all the gas regardless where exactly the encoding is being done.

But the way it is now is more error-prone (except being less usable): if the calling contract somehow screws up correctly encoding the arguments into bytes it would still make a valid `MultiMillerLoop` request. If we take care of encoding it'd be impossible. Not to mention API versioning

Reply:
Supporting @ekovalev care about type safety and @gshep idea of optimised passing by different instances, I'd like to suggest you implement one more type that will be used inside programs for exact Decoding

Reply:
Same for encoding mb?

Reply:
And these will have "Raw" postfix

Reply:
discussed: use current implementation